File: c/shop.c
// Shop using Structures in C
// Author: Clare Tubridy

#include <stdio.h>
#include <string.h>
#include <stdlib.h>

struct Product {
	char* name;
	double price;
};

struct ProductStock {
	struct Product product;
	int quantity;
};

struct Shop {
	double cash;
	struct ProductStock stock[20];
	int index;
};

struct Customer {
	char* name;
	double budget;
	struct ProductStock shoppingList[10];
	int index;
};

// Function to print product details
void printProduct(struct Product p)
{
	printf("PRODUCT NAME: %s \nPRODUCT PRICE: %.2f\n", p.name, p.price);
	printf("-------------\n");
}

// Function to print customer details and shopping list
void printCustomer(struct Customer c)
{
	printf("CUSTOMER NAME: %s \nCUSTOMER BUDGET: %.2f\n", c.name, c.budget);
	printf("-------------\n");
	for(int i = 0; i < c.index; i++)
	{
		printProduct(c.shoppingList[i].product);
		printf("%s ORDERS %d OF ABOVE PRODUCT\n", c.name, c.shoppingList[i].quantity);
		double cost = c.shoppingList[i].quantity * c.shoppingList[i].product.price; 
		printf("The cost to %s will be €%.2f\n\n", c.name, cost);
	}
}

// Function to create and stock the shop with data from a CSV file
struct Shop createAndStockShop(char* filePath)
{
    FILE* fp;
    char line[256];  
    float cash;

    fp = fopen(filePath, "r");
    if (fp == NULL) {
        perror("Error opening file");
        exit(EXIT_FAILURE);
    }

    // Read the first line to get the cash value
    if (fgets(line, sizeof(line), fp) != NULL) {
        cash = atof(line);
        struct Shop s = {cash};

        // Read the remaining lines to get stock information
        while (fgets(line, sizeof(line), fp) != NULL) {
            char *n = strtok(line, ",");
            char *p = strtok(NULL, ",");
            char *q = strtok(NULL, ",");
            int quantity = atoi(q);
            double price = atof(p);
            char *name = malloc(sizeof(char) * 50);
            strcpy(name, n);
            struct Product product = {name, price};
            struct ProductStock stockItem = {product, quantity};
            s.stock[s.index++] = stockItem;
        }

        fclose(fp);
        return s;
    } else {
        fclose(fp);
        perror("Error reading file");
        exit(EXIT_FAILURE);
    }
}

// Function to read customer orders
struct Customer readCustomerOrders(char* filePath) 
{
    FILE* fp;
    char line[256];  
    double budget;

    fp = fopen(filePath, "r");
    if (fp == NULL) {
        perror("Error opening file");
        exit(EXIT_FAILURE);
    }

    // Read the first line to get the customer details
    if (fgets(line, sizeof(line), fp) != NULL) {
        char *n = strtok(line, ",");
        char *b = strtok(NULL, ",");
        budget = atof(b);

        // Allocating memory for the name
        char *name = malloc(strlen(n) + 1);
        strcpy(name, n);

        // Creating the customer struct
        struct Customer c = {name, budget};

        // Read the remaining lines to get the shopping list
        while (fgets(line, sizeof(line), fp) != NULL) {
            char *p = strtok(line, ",");
            char *q = strtok(NULL, ",");
            int quantity = atoi(q);
            char *pname = malloc(sizeof(char) * 50);
            strcpy(pname, p);
            struct Product product = {pname};
            struct ProductStock stockItem = {product, quantity};
            c.shoppingList[c.index++] = stockItem;
        }

        fclose(fp);
        return c;
    } else {
        fclose(fp);
        perror("Error reading file");
        exit(EXIT_FAILURE);
    }
}

// Function to process a transaction
void processTransaction(struct Customer* c, struct Shop* s) {
    double totalCost = 0;
    // Flag to check if any item is in stock
    int itemsInStock = 0; 

    for (int i = 0; i < c->index; i++) {
        int productInStock = 0;

        for (int j = 0; j < s->index; j++) {
            if (strcmp(c->shoppingList[i].product.name, s->stock[j].product.name) == 0) {
                double cost = 0;

                if (c->shoppingList[i].quantity > s->stock[j].quantity) {
                    // Customer ordered more than what is in stock
                    printf("Warning: There are only %d %s in stock. Charging for available quantity.\n", s->stock[j].quantity, c->shoppingList[i].product.name);
                    cost = s->stock[j].quantity * s->stock[j].product.price;
                    // Set stock to 0
                    s->stock[j].quantity = 0; 
                } else {
                    cost = c->shoppingList[i].quantity * s->stock[j].product.price;
                    s->stock[j].quantity -= c->shoppingList[i].quantity;
                }

                totalCost += cost;
                productInStock = 1;
                itemsInStock = 1;
                break;
            }
        }

        if (!productInStock) {
            // Product not found in stock
            printf("Sorry, %s is not in stock. Removing it from the order.\n", c->shoppingList[i].product.name);
            // Remove the item from the order
            c->shoppingList[i] = c->shoppingList[--c->index]; 
            // Recheck the current index as it's replaced with the last item
            i--; 
        }
    }

    if (itemsInStock) {
        if (totalCost > 0 && totalCost <= c->budget) {
            printf("Transaction successful! Customer paid €%.2f.\n", totalCost);
            c->budget -= totalCost;
            s->cash += totalCost;
        } else {
            printf("Transaction cancelled! Insufficient funds.\n");
        }
    } else {
        printf("Transaction cancelled! No items in stock.\n");
    }
}

// Function to print shop details
void printShop(struct Shop s)
{
	printf("Shop has %.2f in cash\n", s.cash);
	for (int i = 0; i < s.index; i++)
	{
		printProduct(s.stock[i].product);
		printf("The shop has %d of the above\n\n", s.stock[i].quantity);
	}
}

// Function for live mode
void liveMode(struct Shop* s) {
    char product_name[50];
    int quantity;

    while (1) {
        printf("\nLive Mode - Enter 'exit' to end the session.\n");

        printf("Enter product name: ");
        scanf("%s", product_name);

        // Check if the user wants to exit
        if (strcmp(product_name, "exit") == 0) {
            break;
        }

        printf("Enter quantity: ");
        scanf("%d", &quantity);

        // Check if the product is in stock
        int found = 0;
        for (int i = 0; i < s->index; i++) {
            if (strcmp(product_name, s->stock[i].product.name) == 0) {
                found = 1;
                double cost = quantity * s->stock[i].product.price;
                printf("The cost for %d %s(s) will be €%.2f\n", quantity, product_name, cost);

                // Process the transaction
                if (cost > 0 && cost <= s->cash && quantity <= s->stock[i].quantity) {
                    printf("Transaction successful! Shop received €%.2f.\n", cost);
                    s->cash += cost;
                    s->stock[i].quantity -= quantity;
                } else {
                    printf("Transaction cancelled! Insufficient funds or out of stock.\n");
                }
                break;
            }
        }

        if (!found) {
            printf("Sorry, %s is not in stock.\n", product_name);
        }
    }

    printf("Exiting live mode.\n");
}

int main(void) 
{
	struct Shop s = createAndStockShop("../stock.csv");
    struct Customer c = readCustomerOrders("../customer.csv");
    //struct Customer c = readCustomerOrders("../customer_no_stock.csv");
	//struct Customer c = readCustomerOrders("../customer_cant_afford.csv");

	printShop(s);
	printCustomer(c);

    processTransaction(&c, &s);

    printShop(s);

    liveMode(&s);

    printShop(s);

    return 0;
}

File: python proc/shop.py
# Procedural Shop
# Author: Clare Tubridy

from dataclasses import dataclass, field
from typing import List
import csv

@dataclass
class Product:
    name: str
    price: float = 0.0

@dataclass 
class ProductStock:
    product: Product
    quantity: int

@dataclass
class Shop:
    cash: float = 0.0
    stock: List[ProductStock] = field(default_factory=list)

@dataclass
class Customer:
    name: str = ""
    budget: float = 0.0
    shopping_list: List[ProductStock] = field(default_factory=list)

def create_and_stock_shop():
     # Create and stock the shop with data from the CSV file
    s = Shop()
    with open('../stock.csv') as csv_file:
        csv_reader = csv.reader(csv_file, delimiter=',')
        first_row = next(csv_reader)
        s.cash = float(first_row[0])
        for row in csv_reader:
            p = Product(row[0], float(row[1]))
            ps = ProductStock(p, float(row[2]))
            s.stock.append(ps)
            #print(ps)
    return s

def calculate_costs(c, s):
    # Calculate costs for each item in the customer's shopping list
    for item in c.shopping_list:
        item.product.price = get_product_price(s, item.product.name)

def process_transaction(c, s):
    # Process a transaction for the customer in the shop
    total_cost = 0
    items_to_remove = []

    for item in c.shopping_list:
        product_price = get_product_price(s, item.product.name)
        if product_price is not None:
            cost = item.quantity * product_price
            total_cost += cost
            update_stock_quantities(s, item.product.name, item.quantity)
        else:
            print(f"Sorry, {item.product.name} is not in stock.")
            items_to_remove.append(item)

    # Remove items that are out of stock from the shopping list
    for item in items_to_remove:
        c.shopping_list.remove(item)
    
    # Check if the customer can afford the transaction
    if total_cost > 0 and total_cost <= c.budget:
        print(f"Transaction successful! {c.name} paid €{total_cost:.2f}.")
        c.budget -= total_cost
        s.cash += total_cost
    else:
        print(f"Transaction cancelled! Insufficient funds.")

def get_product_price(s, product_name):
    # Get the price of a product from the shop's stock
    for item in s.stock:
        if item.product.name == product_name:
            return item.product.price
    return None

def update_stock_quantities(s, product_name, requested_quantity):
    # Update stock quantities based on a customer's order
    for stock_item in s.stock:
        if stock_item.product.name == product_name:
            if requested_quantity <= stock_item.quantity:
                stock_item.quantity -= requested_quantity 
            else:
                print(f"\nSorry, there are only {stock_item.quantity} {product_name} in stock.")
                # Set the stock to zero if the requested quantity exceeds stock
                stock_item.quantity = 0
            break

def read_customer(file_path):
    # Read customer data from a CSV file
    with open(file_path) as csv_file:
        csv_reader = csv.reader(csv_file, delimiter=',')
        first_row = next(csv_reader)
        c = Customer(first_row[0], float(first_row[1]))
        for row in csv_reader:
            name = row[0]
            quantity = float(row[1])
            p = Product(name)
            ps = ProductStock(p, quantity)
            c.shopping_list.append(ps)
        return c 

def live_mode(s):
    # Simulate live shopping mode for the shop
    while True:
        print("\nLive Mode - Enter 'exit' to end the session.")

        # Get product name from the user
        product_name = input("Enter product name: ")

         # Check if the user wants to exit
        if product_name.lower() == 'exit':
            break

        # Get quantity from the user
        try:
            quantity = int(input("Enter quantity: "))
        except ValueError:
            print("Invalid quantity. Please enter a valid number.")
            continue

        # Check if the product is in stock
        product_price = get_product_price(s, product_name)
        if product_price is not None:
            cost = quantity * product_price
            print(f'The cost for {quantity} {product_name}(s) will be €{cost:.2f}')

            # Process the transaction
            if cost > 0 and cost <= s.cash:
                print(f"Transaction successful! Shop received €{cost:.2f}.")
                s.cash += cost
                update_stock_quantities(s, product_name, quantity)
            else:
                print("Transaction cancelled! Insufficient funds.")
        else:
            print(f"Sorry, {product_name} is not in stock.")

    print("Exiting live mode.")

def print_product(p):
    # Print details of a product
    print(f'\nPRODUCT NAME: {p.name} \nPRODUCT PRICE: {p.price}')

def print_customer(c, s):
    # Print details of a customer and their shopping list
    print(f'CUSTOMER NAME: {c.name} \nCUSTOMER BUDGET: {c.budget}')
    for item in c.shopping_list:
        print_product(item.product)
        
        product_price = get_product_price(s, item.product.name)
        if product_price is not None:
            print(f'{c.name} ORDERS {item.quantity} OF ABOVE PRODUCT')
            cost = item.quantity * item.product.price
            print(f'The cost to {c.name} will be €{cost:.2f}')
        else:
            print(f"Sorry, {item.product.name} is not in stock.")
        
def print_shop(s):
    # Print details of the shop, including cash and stock
    print(f'\nShop has {s.cash:.2f} in cash')
    for item in s.stock:
        print_product(item.product)
        print(f'The Shop has {item.quantity} of the above')


# Example usage
s = create_and_stock_shop()
#c = read_customer("../customer.csv")
#c = read_customer("../customer_no_stock.csv")
c = read_customer("../customer_cant_afford.csv")
#calculate_costs(c,s)
#print_customer(c,s)
#process_transaction(c,s)
live_mode(s)
print_shop(s)